### ISSUE TITLE
Wrong lighting ogre-mesh rviz 1.11.3 and OGRE 1.8.1 ROS-Indigo

### ISSUE THREAD

**Issue by TorstenWK** [ros-visualization/rviz#830]
Hi All,

 I wrote a rviz-plugin which loads and display a ogre mesh (a vehicle to be more exact). In ROS hydro everything is the more or the less fine, the mesh is drawn as it should and also the lighting is correct.

If i change nothing in the plugin and only change the ROS version to indigo, some really strange things happen:

1. The lighting of the mesh seams to be broken, nearly all surfaces are drawn black.
2. The mesh seams to be drawn twice. I tried to add a GLSL shader program which adds a simple phong-shading to some surfaces / materials. Despite of that the model and view matrix will not get passed correctly to the shader (i used the auto params) i got some heavy z-fighting on surfaces which definitely are not modeled double sided.

What i already tried / did:

1. Double checked normals, they are ok.
2. Viewed the model with the Ogre Max Viewer, which also uses OGRE Version 1.8.1 (same as the ROS indigo) - Lighting is correct!
3. Wrote a simple test program which not uses ROS. - Lighting is correct!
4. Updated mesh with the Update Utility of OGRE.
5. Tested another mesh - Same issue!
6. Programmatically changed Face Culling, Light, various material values of the mesh, use own GLSL shader (see above) Nothing changed. ( Set ambient light in scene-manager, add additional light to scene ... )
7. Exported a simple cube from Blender (start scene cube) 2.69.0 with blender2ogre-0.6.0 and tested this mesh in both ROS versions. Same here, Cube is black in ROS indigo and white (correct lighting) in ROS hydro. Both with identical rviz plugin.
8. Cloned newest rviz version (indigo-devel-branch) to ROS-hydro version so rviz 1.11.3 will run under ROS-hydro, everything works well! I also tested the old rviz version 1.10.18 (hydro-devel-branch) under ROS-indigo, lighting still broken. So it seems that the new Ogre-Version 1.8.1 is the cause of this problem, or maybe a not correctly adapted rviz.

What seems to work:

1. Emissive materials seems to work in ROS indigo. The cube will be red if emissive set to 1 0 0 1 in ogre-material file.

Did someone of you encounter the same problems drawing OGRE Meshes in ROS-indigo-rviz?

Mesh in ROS Hydro (RViz 0.10.18, OGRE 1.7.3):

![ros_hydro](https://cloud.githubusercontent.com/assets/9879001/5142015/6cbba288-717d-11e4-8dc5-53bcad48b81a.png)

Mesh in ROS Indigo (RViz 0.11.3, OGRE 1.8.1):

![rosindigo](https://cloud.githubusercontent.com/assets/9879001/5142018/72bbf638-717d-11e4-9050-9cdfc3c1c6ea.png)

Post in ROS answers:

http://answers.ros.org/question/197839/ros-indigo-rviz-ogre-mesh-wrong-lighting/

Maybe you can help me with this.

With best Regards,

Torsten

**Comment by VictorLamoine** [comment]
I had a similar (not sure if the same) issue; the material description in `dae` was not correct, I replaced
```
<ambient>
  <color sid="ambient">0 0 0 1</color>
</ambient>
```
with
```
<ambient>
  <color sid="ambient">0.1 0.1 0.1 1</color>
</ambient>
```
And things got better.
Hope this helps

**Comment by wjwwood** [comment]
@VictorLamoine if what you described is true, perhaps https://github.com/ros-visualization/rviz/pull/837 fixed this issue too. It's still on my list to reproduce this at which point I can check to see if that pr had any effect.

**Comment by wjwwood** [comment]
@TorstenWK I went ahead and tried to reproduce what you described using your cube with rviz versions 1.10.18, 1.11.4, and the indigo-devel branch. In all cases this was done on my trusty machine and with ogre 1.8 installed:

```
% dpkg -l | grep ogre
ii  libogre-1.8-dev                                       1.8.1+dfsg-0ubuntu3                                 amd64        3D Object-Oriented Graphics Rendering Engine (development files)
ii  libogre-1.8.0:amd64                                   1.8.1+dfsg-0ubuntu3                                 amd64        3D Object-Oriented Graphics Rendering Engine (libraries)
```

And in all cases I got a white cube, not the black cube as you described. Just to confirm you expected a black cube with the files you linked from google drive on Indigo's rviz?

**Comment by wjwwood** [comment]
I also tried `1.11.3.` with the same result. I put together this test package as a more end to end check for the bug:

https://github.com/wjwwood/test_mesh_urdf_rviz

Can you take a look at that package and see if there is anything about it that you think would prevent me from reproducing the problem?

**Comment by juftho** [comment]
we still have the problem with the car mesh in ROS indigo.
I tried your example in rviz and ogremeshy: rviz shows a white box, ogremeshy a black/grey box.
(rviz 1.11.3, ROS indigo, Ogre 1.8.1, OpenGl 4.4).
Is the problem somehow related to http://www.ogre3d.org/forums/viewtopic.php?f=2&t=70493 ?

**Comment by wjwwood** [comment]
Interesting, have you tried it with Ogre 1.9? They have reported fixed it, though since they recently moved to JIRA I'm unable to figure out which version was fixed in or even what the patch was.

**Comment by juftho** [comment]
Thanks for your help again. I just checked the ogre-forum - the bug was fixed here
https://bitbucket.org/sinbad/ogre/commits/d45e64c81b4c15f21efbcb8d5c03eef79f08c23d#general-comments

It seems that the change is already included in the Ubuntu Package for 14.04:
http://packages.ubuntu.com/source/trusty/ogre-1.8

I've installed ogre+ROS via Ubuntu Package manager and can't easily switch to a different ogre version (or a version compiled by me). Is there a handy way to try a newer ogre 1.8.x version with rviz?

btw, my ogremeshy runs with its own ogre lib which is v1.9. Since the mesh is loaded correctly there, it's likely a bug in ogre which was fixed in a later version.

**Comment by juftho** [comment]
Downloading and compiling the current Ogre 1.8-Branch solved the problem! So it is a ogre bug which is fixed in the current 1.8-branch, but not in the 1.8-version provided by Ubuntu Package Manager.

However, it would be great to have the Ubuntu-Packages of ROS available with Ogre-1.9 instead of 1.8. The Ogre-1.9-Package is already available in Ubuntu 14.04.

@wjwwood: Thanks for your help!

**Comment by wjwwood** [comment]
@TorstenWK can you give it a try with the latest Ogre 1.8 (from source) or with Ogre 1.9? I would guess that this bug fix will address you issue as well.

Unfortunately our current plan is to stick with Ogre 1.8 on Ubuntu Trusty for both Indigo and Jade. For Jade on Ubuntu Utopic and Vivid we will be using Ogre 1.9. We're doing this on Trusty as to not conflict with the Gazebo packages which use Ogre 1.8 as well.

The best bet at this point for fixing it in Indigo on Trusty is finding a work around based on the bug fix in the input meshes or pushing for the Debian packagers or the OGRE developers to get the fix back ported into Trusty.

I'm going to close this for now since there is nothing we can do in rviz to resolve it (as far as I can see). If anyone has more feedback or disagrees please comment here and I can reopen the issue if necessary.
